lib/AST/GenericEnvironment.cpp

```diff
@@ -53,6 +53,9 @@ GenericEnvironment::GenericEnvironment(
     // FIXME: If multiple generic parameters map to the same archetype,
     // the reverse mapping order is not deterministic.
   }
+
+  // Make sure this generic environment gets destroyed.
+  signature->getASTContext().addDestructorCleanup(*this);
```

Comment:
This makes me unhappy; it seems like it could greatly increase the number of destructor cleanups we need. Since the mapping won't change during the lifetime of the generic environment, can we use some kind of sorted list and tail-allocate the storage instead?